File: sorts/quick_sort/python3/quick_sort.py
from typing import List


def quick_sort(collection: List, reverse=False) -> List:
	"""A pure Python implementation of quick sort algorithm.
    :param collection: a mutable collection of comparable items
    :return: the same collection ordered by ascending"""
	if len(collection) < 2:
		return collection
	pivot = collection.pop()
	greater: List[int] = []  # All elements from collection which greater pivot
	lesser: List[int] = []  # All elements from collection which less or equal pivot
	for element in collection:
		(greater if element > pivot else lesser).append(element)

	if (not reverse):
		return quick_sort(lesser) + [pivot] + quick_sort(greater)
	else:
		return quick_sort(greater) + [pivot] + quick_sort(lesser)

File: sorts/quick_sort/swift/quick_sort.swift
func quickSort<T: Comparable>(_ array: [T]) -> [T] {
    guard array.count > 1 else { return array }

    let pivot = array[Int.random(in: 0...array.count-1)]
    // sub-array of all the elements less than the pivot
    let less = array.filter {$0 < pivot}
    // sub-array of all the elements equal to the pivot
    let equal = array.filter {$0 == pivot}
    // sub-array of all the elements greater than the pivot
    let greater = array.filter {$0 > pivot}
    return quickSort(less) + equal + quickSort(greater)
}
